hudi-common/src/main/java/org/apache/hudi/common/data/HoodieList.java

```diff
@@ -132,6 +134,26 @@ public long count() {
     return HoodieList.of(new ArrayList<>(new HashSet<>(listData)));
   }
 
+  @Override
+  public <O> HoodieData<T> distinctWithKey(SerializableFunction<T, O> keyGetter, int parallelism) {
+    Set<O> set = listData.stream().map(i -> throwingMapWrapper(keyGetter).apply(i)).collect(Collectors.toSet());
+    List<T> distinctList = new LinkedList<>();
+    listData.forEach(x -> {
+      if (set.contains(throwingMapWrapper(keyGetter).apply(x))) {
+        distinctList.add(x);
+      }
+    });
+    return HoodieList.of(distinctList);
```

Comment:
It seems that the result can still have multiple objects with the same key.  But the `distinct` should only give back one object per key?

Reply:
fixed with new logic and filing a ticket for Flink/Java fix